### DISCUSSION TITLE
How to Embed pymadcad in PyQt5?

### DISCUSSION THREAD

**Discussion by Mehrshad-Ghorbanzadeh** [jimy-byerley/pymadcad#11] [Q&A] [answered]
Hello and thanks for your extraordinary effort. Is it possible to define a widget for pymadcad in pyqt5 similar to matplotlib and use it in an application?

**Comment by jimy-byerley** [comment, marked as answer]
Hello and welcome there :) 
Sure there is !
It's called `View` which is a widget that renders a `Scene`. which is typically what `madcad.show` does

-  [madcad.rendering.View](https://pymadcad.readthedocs.io/en/latest/reference/rendering.html#madcad.rendering.View)  is a `QWidget` that can be integrated in pretty any Qt stuff. It even support touchscreen events ;)
- [madcad.rendering.Scene](https://pymadcad.readthedocs.io/en/latest/reference/rendering.html#madcad.rendering.Scene) is a pure python object that manage the openGL stuff ( display object buffers and so on) and can be shared across instances of `View`. So the same scene exactly can be seen under different point of view.

So here is a typical example:
```python
# create application and window, normal stuff
app = QApplication(sys.argv)
window = QWidget()

# integrate a view on the 3D scene
view = View(Scene(objs_that_you_pass_to_show), parent=window)

# start everything, normal stuff
window.show()
app.exec()
```

**Reply by Mehrshad-Ghorbanzadeh** [reply]
Thanks for your quick and comprehensive response. This worked and I could use it as a widget. There is another problem, I wrote the following code and worked with your method, but the display setting does not work correctly. This is not a verticalLayout problem because QWidget is displayed correctly. Thank you for your help in this case as well.

```python
from madcad.rendering import View
from madcad.rendering import Scene
from madcad.mathutils import *
from madcad.generation import revolution
from madcad.generation import web
from madcad.primitives import Circle
from madcad.rendering import show
from madcad import settings
import sys
from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QApplication, QMainWindow, QDialog, QWidget
from PyQt5.QtCore import Qt, QPoint, QEvent


class Ui_Dialog(object):
    def setupUi(self, Dialog):
        Dialog.setObjectName("Dialog")
        Dialog.resize(400, 300)
        self.verticalLayout = QtWidgets.QVBoxLayout(Dialog)
        self.verticalLayout.setObjectName("verticalLayout")
        self.widget = QtWidgets.QWidget(Dialog)
        self.widget.setObjectName("widget")
        self.verticalLayout.addWidget(self.widget)
        self.retranslateUi(Dialog)
        QtCore.QMetaObject.connectSlotsByName(Dialog)
    def retranslateUi(self, Dialog):
        _translate = QtCore.QCoreApplication.translate
        Dialog.setWindowTitle(_translate("Dialog", "Dialog"))

class MainGui(QDialog):
    def __init__(self):
        super().__init__()
        self.uiMainWin = Ui_Dialog()
        self.uiMainWin.setupUi(self)
        self.show()
        ############################################################
        self.objs = self.make_objects()
        ############################################################
        #app  = self.make_app()
        ############################################################
        view = self.make_view()
    #######################
    def make_objects(self):
        O = vec3(0)
        A = vec3(2,0,0)
        B = vec3(1,2,0)
        C = vec3(0,2,0)
        part = revolution( radians(180), (O,Z), web( Circle( (A,Y), 0.5) ) )
        objs = [part]
        if isinstance(objs, list):
            objs = dict(enumerate([objs]))
        return objs
    #######################
    def make_app(self):
        app = QApplication.instance()
        if not app:
            QApplication.setAttribute(Qt.AA_ShareOpenGLContexts, True)
            app = QApplication(sys.argv)
        return app
    #######################
    def make_view(self):
        if settings.display['system_theme']:
            settings.use_qt_colors()
        ###
        self.view = View(Scene(self.objs, options=None), parent=self.uiMainWin.widget)
        self.view.show()
        ###
        interest = None
        if not interest:
            interest = self.view.scene.box()
        self.view.center()
        self.view.adjust()
    #######################

if __name__=="__main__":
    #QApplication.setAttribute(Qt.AA_ShareOpenGLContexts, True)
    app = QApplication(sys. argv)
    wi = MainGui()
    wi.show()
    sys.exit(app.exec_())
```

**Reply by jimy-byerley** [reply]
Well I'm not sure what exactly you are expecting for.
The view widget is small, it's pretty normal:  the `View` widget is parented to `self.uiMainWin.widget` which is just a simple `QWidget` with no layout, hence it doesn't resize its children. So the `View` has a default size (which depend on Qt).
Putting the view inside a widget with layout, or setting it a size will change that.

Was that what you found wrong in your screenshot ?

**Reply by Mehrshad-Ghorbanzadeh** [reply]
Thank you so much. I change the addwidget of verticalLayout. and it works. First, View(Sense) must be done and then addwidget.verticalLayout. This was occured because I used qtdesigner. I hope you are successful and thank you for your help.
This is the new code : 
from madcad.rendering import View
from madcad.rendering import Scene
from madcad.mathutils import *
from madcad.generation import revolution
from madcad.generation import web
from madcad.primitives import Circle
from madcad.rendering import show
from madcad import settings
import sys
from PyQt5 import QtCore, QtGui, QtWidgets
from PyQt5.QtWidgets import QApplication, QMainWindow, QDialog, QWidget
from PyQt5.QtCore import Qt, QPoint, QEvent


class Ui_Dialog(object):
    def setupUi(self, Dialog):
        Dialog.setObjectName("Dialog")
        Dialog.resize(400, 300)
        self.verticalLayout = QtWidgets.QVBoxLayout(Dialog)
        self.verticalLayout.setObjectName("verticalLayout")
        self.widget = QtWidgets.QWidget(Dialog)
        self.widget.setObjectName("widget")
        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
        #self.verticalLayout.addWidget(self.widget)
        self.retranslateUi(Dialog)
        QtCore.QMetaObject.connectSlotsByName(Dialog)
    def retranslateUi(self, Dialog):
        _translate = QtCore.QCoreApplication.translate
        Dialog.setWindowTitle(_translate("Dialog", "Dialog"))

class MainGui(QDialog):
    def __init__(self):
        super().__init__()
        self.uiMainWin = Ui_Dialog()
        self.uiMainWin.setupUi(self)
        self.show()
        ############################################################
        self.objs = self.make_objects()
        ############################################################
        #app  = self.make_app()
        ############################################################
        view = self.make_view()
    #######################
    def make_objects(self):
        O = vec3(0)
        A = vec3(2,0,0)
        B = vec3(1,2,0)
        C = vec3(0,2,0)
        part = revolution( radians(180), (O,Z), web( Circle( (A,Y), 0.5) ) )
        objs = [part]
        if isinstance(objs, list):
            objs = dict(enumerate([objs]))
        return objs
    #######################
    def make_app(self):
        app = QApplication.instance()
        if not app:
            QApplication.setAttribute(Qt.AA_ShareOpenGLContexts, True)
            app = QApplication(sys.argv)
        return app
    #######################
    def make_view(self):
        if settings.display['system_theme']:
            settings.use_qt_colors()
        ###
        self.view = View(Scene(self.objs, options=None), parent=self.uiMainWin.widget)
        self.uiMainWin.verticalLayout.addWidget(self.view)


        self.view.show()
        ###
        interest = None
        if not interest:
            interest = self.view.scene.box()
        self.view.center()
        self.view.adjust()
    #######################

if __name__=="__main__":
    #QApplication.setAttribute(Qt.AA_ShareOpenGLContexts, True)
    app = QApplication(sys. argv)
    wi = MainGui()
    wi.show()
    sys.exit(app.exec_())

**Reply by jimy-byerley** [reply]
haha, I personnaly do not like QtDesigner so much for such reasons
But I'm glad you got it to work ! Good luck for your project :+1:

**Comment by Mehrshad-Ghorbanzadeh** [comment]
This is the result.
![1](https://user-images.githubusercontent.com/53825204/145192090-2cfdceb9-9a2e-4d79-b10b-b3a2dffe9834.PNG)

**Comment by Mehrshad-Ghorbanzadeh** [comment]
![1](https://user-images.githubusercontent.com/53825204/145264551-e9669cab-7519-4474-9bbc-cdf67c9ae736.PNG)
